src/testdrive/src/action/loki.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::{collections::HashMap, time::Duration};
+
+use anyhow::{anyhow, Context};
+use async_trait::async_trait;
+use chrono::Utc;
+use mz_ore::retry::Retry;
+use serde::Serialize;
+use url::Url;
+
+use crate::action::{Action, ControlFlow, State};
+use crate::parser::BuiltinCommand;
+
+pub struct IngestAction {
+    /// Labels to be added to the ingested lines.
+    labels: HashMap<String, String>,
+    /// The lines to be ingested.
+    lines: Vec<String>,
+}
+
+pub fn build_ingest(cmd: BuiltinCommand) -> Result<IngestAction, anyhow::Error> {
+    let labels = cmd.args.into_iter().collect();
+    Ok(IngestAction {
+        labels,
+        lines: cmd.input,
+    })
+}
+
+#[async_trait]
+impl Action for IngestAction {
+    async fn undo(&self, _state: &mut State) -> Result<(), anyhow::Error> {
+        Ok(())
+    }
+
+    async fn redo(&self, state: &mut State) -> Result<ControlFlow, anyhow::Error> {
+        state
+            .loki_client
+            .push_logs(&self.labels, &self.lines)
+            .await?;
+        Ok(ControlFlow::Continue)
+    }
+}
+
+pub struct Client {
+    url: Url,
+    client: reqwest::Client,
+    timeout: Duration,
+}
```

Comment:
We could extract this at some point and merge it with the `LokiSourceReader::query`. No need for now, though.